### PR TITLE
Insights refresh improved

### PR DESCRIPTION
**Pull request by Twixes**
## Changes

The UX of insight refreshes has been weird since it's been added. For instance, the option used to weirdly disappear in a puff of smoke after being used. It also was unclear when was it shown to begin with, and it didn't update the time of last recomputing when the tab was left open for longer.
This is now fixed for Funnels and for other insights.

Before:

After:

## Checklist

- [x] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] ~~Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious~~
- [x] New/changed UI is decent on smartphones (viewport width around 360px)

### CODE REVIEW

File: posthog/decorators.py [hunk 1]
```diff
@@ -25,37 +34,45 @@ class CacheType(str, Enum):
...
+
+        filter = get_filter(request=request, team=team)
+        cache_key = generate_cache_key("{}_{}".format(filter.toJSON(), team.pk))
+
+        # return cached result if possible
+        if not should_refresh(request):
+            cached_result_package = get_safe_cache(cache_key)
+            if cached_result_package and cached_result_package.get("result"):
+                cached_result_package["is_cached"] = True
+                return cached_result_package
```
Comment (alexkim205) [inline posthog/decorators.py:60]: Do we know if `last_refresh` is always included in this result package for cached responses?

Comment (Twixes) [reply posthog/decorators.py:60]: We do, because we set it a few lines below.

**Comment by alexkim205** [comment]
lgtm! Tangentially related but backend changes here are a prereq for #5522

**Review by alexkim205 — APPROVED** [review]

File: posthog/decorators.py [hunk 2]
```diff
@@ -25,37 +34,45 @@ class CacheType(str, Enum):
...
+T = TypeVar("T", bound=ResultPackage)
+U = TypeVar("U", bound=GenericViewSet)
+
+
+def cached_function(f: Callable[[U, Request], T]) -> Callable[[U, Request], T]:
+    @wraps(f)
+    def wrapper(self, request) -> T:
+        # prepare caching params
+        team = cast(User, request.user).team
+        filter = None
```
Comment (alexkim205) [inline posthog/decorators.py:48]: Not an issue with your changes, but an artifact of today's code, but can we remove this local `filter` here? It doesn't look like it's being used anywhere.

File: posthog/decorators.py [hunk 3]
```diff
@@ -1,13 +1,22 @@
-from datetime import datetime
 from enum import Enum
 from functools import wraps
-from typing import Callable, Dict, List, Union, cast
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    List,
+    Optional,
```
Comment (alexkim205) [inline posthog/decorators.py:8]: I think this Optional is not used

File: posthog/decorators.py (hunk 2 above)
Comment (Twixes) [reply posthog/decorators.py:48]: Indeed

**Comment by Twixes** [comment]
Adjusted the code @alexkim205. I'll go ahead and merge.

**Pull request merged**